src/module_5/module5.md

```diff
@@ -0,0 +1,5880 @@
+    }
+
+    .dataframe tbody tr th {
+        vertical-align: top;
+    }
+
+    .dataframe thead th {
+        text-align: right;
+    }
+</style>
+<table border="1" class="dataframe">
+  <thead>
+    <tr style="text-align: right;">
+      <th></th>
+      <th>logloss</th>
+      <th>custom_eval_metric</th>
+      <th>n_trees</th>
+      <th>execution_date</th>
+      <th>diff_ch_sp500_baseline</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <th>0</th>
+      <td>0.716178</td>
+      <td>0.364069</td>
+      <td>19</td>
+      <td>2006-06-30</td>
+      <td>0.049213</td>
+    </tr>
+    <tr>
+      <th>1</th>
+      <td>0.704118</td>
+      <td>-0.029699</td>
+      <td>19</td>
+      <td>2006-09-30</td>
+      <td>0.067796</td>
+    </tr>
+    <tr>
+      <th>2</th>
+      <td>0.711382</td>
+      <td>-0.001069</td>
+      <td>19</td>
+      <td>2006-12-30</td>
+      <td>0.068473</td>
+    </tr>
+    <tr>
+      <th>3</th>
+      <td>0.701398</td>
+      <td>-0.066858</td>
+      <td>19</td>
+      <td>2007-03-31</td>
+      <td>0.048029</td>
+    </tr>
+    <tr>
+      <th>4</th>
+      <td>0.695110</td>
+      <td>-0.030262</td>
+      <td>19</td>
+      <td>2007-06-30</td>
+      <td>0.077166</td>
+    </tr>
+  </tbody>
+</table>
+</div>
+
+
+
+
+```python
+(
+    ggplot(
+        test_results_final_tree[test_results_final_tree["custom_eval_metric"] < 10],
+        aes(x="execution_date"),
+    )
+    + geom_point(aes(y="custom_eval_metric"), colour="blue")
+    + geom_point(aes(y="diff_ch_sp500_baseline"), colour="red")
+    + theme(axis_text_x=element_text(angle=90, vjust=0.5, hjust=1))
+)
+```
+
+
+    
+![png](module5_files/module5_112_0.png)
+    
+
+
+
+
+
+    <Figure Size: (640 x 480)>
+
+
+
+
+```python
+print(f"Mean for model: {test_results_final_tree['custom_eval_metric'].mean()}")
+print(
+    f"Mean for baseline: {test_results_final_tree['diff_ch_sp500_baseline'].mean()}\n"
+)
+
+print(f"Median for model: {test_results_final_tree['custom_eval_metric'].median()}")
+print(
+    f"Median for baseline: {test_results_final_tree['diff_ch_sp500_baseline'].median()}\n"
+)
+```
+
+    Mean for model: 0.09278869642857143
+    Mean for baseline: 0.022159133577893696
+    
+    Median for model: 0.011363999999999999
+    Median for baseline: 0.015525563344158869
+    
+
+
+## Conclusions
+
+**Mean and Median Analysis: Model vs. Baseline**.
+The mean value of your model stands at approximately 9.28%, significantly higher than the baseline's mean of about 2.22%. This indicates that on average, your model predicts considerably higher returns compared to the baseline. However, when looking at the median values, a different picture emerges. The model's median is around 1.14%, lower than the baseline's 1.55%. This suggests that while the model excels in generating higher average returns, its performance across various scenarios is less consistent, as the median is more resistant to outliers than the mean.
```

Comment:
A possible explanation of this is that if you think about it the model is giving high probabilities to companies which are quite similar (or in other words, the returns are very correlated). This means that either all the companies go well or none of them do (so very high or low returns). When doing classic portfolios, one tries to find companies which a) good returns b) which are uncorrelated (check out https://en.wikipedia.org/wiki/Modern_portfolio_theory), so it would be quite interesting to think of a regularization terms which penalized correlated things